Support Arm dynamic linking

DrXiao opened the pull request:
The proposed changes enhance shecc to generate dynamically linked executables. When the `--dynlink` flag is specified, shecc produces sections such as `.plt` and `.got` for the compiled programs, allowing the executables to leverage the ELF interpreter and the GNU C library to run.

This pull request is still a work in progress due to the following incomplete tasks:
* [x] Fix the potential issues. (The bootstrapping process still fails for dynamically linked shecc.)
* [x] Improve code quality and commit messages.
* [x] Make the dynamically linked shecc run the test suites.
* [x] Improve `README.md` to describe dynamic linking.
* [x] Enhance GitHub workflows to verify the dynamically linked shecc.
* [x] Validate the proposed changes on an Arm machine such as a BeagleBone Black or Raspberry Pi.
* [x] Refine `c.c` and `c.h` to avoid duplications
* [x] Update and separate the snapshots for static linking and dynamic linking.
* [x] Add required documentation

Updated usage: (10/11 11:43 updated)
```shell
# Perform bootstrapping process for the dynamically linked shecc.
$ make DYNLINK=1

# Add '--dynlink' to generate dynamically linked executable.
$ shecc [-o output] [+m] [--dump-ir] [--no-libc] [--dynlink] <input.c>

# Run the generated executable by given the elf interpreter prefix.
$ qemu-arm -L /usr/arm-linux-gnueabihf/ <executable>
```

DrXiao commented:
FWIW, I disassemble the `test` executable:

<details>
<summary> test.asm </summary>

```diff
$ arm-linux-gnueabi-objdump -d test

test:     file format elf32-littlearm


Disassembly of section .text:

000100b4 <.text>:
   100b4:	e3a0b000 	mov	fp, #0
   100b8:	e3a0e000 	mov	lr, #0
   100bc:	e49d1004 	pop	{r1}		@ (ldr r1, [sp], #4)
   100c0:	e1a0200d 	mov	r2, sp
   100c4:	e52d2004 	push	{r2}		@ (str r2, [sp, #-4]!)
   100c8:	e52d0004 	push	{r0}		@ (str r0, [sp, #-4]!)
   100cc:	e3a0c000 	mov	ip, #0
   100d0:	e52dc004 	push	{ip}		@ (str ip, [sp, #-4]!)
   100d4:	e30000ec 	movw	r0, #236	@ 0xec
   100d8:	e3400001 	movt	r0, #1
   100dc:	e3a03000 	mov	r3, #0
   100e0:	eb000067 	bl	0x10284
   100e4:	e3a0007f 	mov	r0, #127	@ 0x7f
   100e8:	eb000005 	bl	0x10104
   100ec:	e1a09000 	mov	r9, r0
   100f0:	e1a0a001 	mov	sl, r1
   100f4:	e3008004 	movw	r8, #4
   100f8:	e3408000 	movt	r8, #0
   100fc:	e04dd008 	sub	sp, sp, r8
   10100:	e1a0c00d 	mov	ip, sp
   10104:	eb000005 	bl	0x10120
   10108:	e3008004 	movw	r8, #4
   1010c:	e3408000 	movt	r8, #0
   10110:	e08dd008 	add	sp, sp, r8
   10114:	e1a00000 	nop			@ (mov r0, r0)
   10118:	e3a07001 	mov	r7, #1
   1011c:	ef000000 	svc	0x00000000
   10120:	e1a00009 	mov	r0, r9
   10124:	e1a0100a 	mov	r1, sl
   10128:	eaffffff 	b	0x1012c
   1012c:	e50de004 	str	lr, [sp, #-4]
   10130:	e3008044 	movw	r8, #68	@ 0x44
   10134:	e3408000 	movt	r8, #0
   10138:	e04dd008 	sub	sp, sp, r8
   1013c:	e3000224 	movw	r0, #548	@ 0x224
   10140:	e3400001 	movt	r0, #1
   10144:	e3a01001 	mov	r1, #1
   10148:	e3a02002 	mov	r2, #2
   1014c:	e3a03003 	mov	r3, #3
   10150:	e58d0004 	str	r0, [sp, #4]
   10154:	e58d1008 	str	r1, [sp, #8]
   10158:	e58d200c 	str	r2, [sp, #12]
   1015c:	e58d3010 	str	r3, [sp, #16]
   10160:	e59d0004 	ldr	r0, [sp, #4]
   10164:	e59d1008 	ldr	r1, [sp, #8]
   10168:	e59d200c 	ldr	r2, [sp, #12]
   1016c:	e59d3010 	ldr	r3, [sp, #16]
+  10170:	eb000046 	bl	0x10290
   10174:	e300022e 	movw	r0, #558	@ 0x22e
   10178:	e3400001 	movt	r0, #1
   1017c:	e3a01001 	mov	r1, #1
   10180:	e3a02002 	mov	r2, #2
   10184:	e3a03003 	mov	r3, #3
   10188:	e3a04004 	mov	r4, #4
   1018c:	e58d0014 	str	r0, [sp, #20]
   10190:	e58d1018 	str	r1, [sp, #24]
   10194:	e58d201c 	str	r2, [sp, #28]
   10198:	e58d3020 	str	r3, [sp, #32]
   1019c:	e58d4024 	str	r4, [sp, #36]	@ 0x24
   101a0:	e59d0014 	ldr	r0, [sp, #20]
   101a4:	e59d1018 	ldr	r1, [sp, #24]
   101a8:	e59d201c 	ldr	r2, [sp, #28]
   101ac:	e59d3020 	ldr	r3, [sp, #32]
   101b0:	e59d4024 	ldr	r4, [sp, #36]	@ 0x24
+  101b4:	eb000035 	bl	0x10290
   101b8:	e300023b 	movw	r0, #571	@ 0x23b
   101bc:	e3400001 	movt	r0, #1
   101c0:	e3a01001 	mov	r1, #1
   101c4:	e3a02002 	mov	r2, #2
   101c8:	e3a03003 	mov	r3, #3
   101cc:	e3a04004 	mov	r4, #4
   101d0:	e3a05005 	mov	r5, #5
   101d4:	e58d0028 	str	r0, [sp, #40]	@ 0x28
   101d8:	e58d102c 	str	r1, [sp, #44]	@ 0x2c
   101dc:	e58d2030 	str	r2, [sp, #48]	@ 0x30
   101e0:	e58d3034 	str	r3, [sp, #52]	@ 0x34
   101e4:	e58d4038 	str	r4, [sp, #56]	@ 0x38
   101e8:	e58d503c 	str	r5, [sp, #60]	@ 0x3c
   101ec:	e59d0028 	ldr	r0, [sp, #40]	@ 0x28
   101f0:	e59d102c 	ldr	r1, [sp, #44]	@ 0x2c
   101f4:	e59d2030 	ldr	r2, [sp, #48]	@ 0x30
   101f8:	e59d3034 	ldr	r3, [sp, #52]	@ 0x34
   101fc:	e59d4038 	ldr	r4, [sp, #56]	@ 0x38
   10200:	e59d503c 	ldr	r5, [sp, #60]	@ 0x3c
+  10204:	eb000021 	bl	0x10290
   10208:	e3a00000 	mov	r0, #0
   1020c:	e1a00000 	nop			@ (mov r0, r0)
   10210:	e3008044 	movw	r8, #68	@ 0x44
   10214:	e3408000 	movt	r8, #0
   10218:	e08dd008 	add	sp, sp, r8
   1021c:	e51de004 	ldr	lr, [sp, #-4]
   10220:	e12fff3e 	blx	lr

Disassembly of section .plt:

00010270 <.plt>:
   10270:	e52de004 	push	{lr}		@ (str lr, [sp, #-4]!)
   10274:	e300a2a4 	movw	sl, #676	@ 0x2a4
   10278:	e340a001 	movt	sl, #1
   1027c:	e1a0e00a 	mov	lr, sl
   10280:	e59ef000 	ldr	pc, [lr]
   10284:	e300c2a8 	movw	ip, #680	@ 0x2a8
   10288:	e340c001 	movt	ip, #1
   1028c:	e59cf000 	ldr	pc, [ip]
+  10290:	e300c2ac 	movw	ip, #684	@ 0x2ac
   10294:	e340c001 	movt	ip, #1
   10298:	e59cf000 	ldr	pc, [ip]
```

</details>

`0x10290` is the starting point of `printf@plt`. However, in the `text` section, there are three places using `bl` instruction to call `printf()`, and each of these places has several `str` and `ldr` instructions to manipulate the stack beforehand.

jserv commented on Makefile:
```diff
@@ -36,6 +36,15 @@ STAGE0 := shecc
+STAGE0_FLAGS ?= --dump-ir
+STAGE1_FLAGS ?=
+ifeq ($(DYNLINK),1)
+BUILTIN_LIBC := dyn-c.c
```
Can you avoid `dyn-c.c`?

jserv commented:
I would like to ask @lecopzer  for reviewing.

jserv commented:
With the changes below, I can proceed with stage 1 compilation via 'make DYNLINK=1':
```diff
diff --git a/src/arm-codegen.c b/src/arm-codegen.c
index 6fb1079..39f3396 100644
--- a/src/arm-codegen.c
+++ b/src/arm-codegen.c
@@ -138,7 +138,8 @@ void cfg_flatten(void)
     func_t *func;
 
     if (dynlink)
-        elf_offset = 108; /* offset of start + exit in codegen */
+        elf_offset =
+            92; /* offset of dynamic linking setup + global init call */
     else {
         func = find_func("__syscall");
         func->bbs->elf_offset = 44; /* offset of start + exit in codegen */
@@ -154,7 +155,7 @@ void cfg_flatten(void)
 
     /* prepare 'argc' and 'argv', then proceed to 'main' function */
     if (dynlink)
-        elf_offset += 20;
+        elf_offset += 12; /* 3 insns: restore r0/r1 from r9/r10, bl to main */
     else
         elf_offset += 32; /* 6 insns for main call + 2 for exit */
 
@@ -481,8 +482,18 @@ void code_generate(void)
         emit(__push_reg(__AL, __r0));
         emit(__mov_i(__AL, __r12, 0));
         emit(__push_reg(__AL, __r12));
-        emit(__movw(__AL, __r0, elf_code_start + 56));
-        emit(__movt(__AL, __r0, elf_code_start + 56));
+        /* Pass the address of our main wrapper function
+         * After these movw/movt, we have:
+         * - mov r3, #0 (4 bytes)
+         * - bl to PLT (4 bytes)
+         * - mov r0, #127 (4 bytes)
+         * - bl +28 (4 bytes)
+         * - main wrapper starts here
+         * Total offset = current + 8 (movw+movt) + 16 = current + 24
+         */
+        int main_wrapper_offset = elf_code->size + 24;
+        emit(__movw(__AL, __r0, elf_code_start + main_wrapper_offset));
+        emit(__movt(__AL, __r0, elf_code_start + main_wrapper_offset));
         emit(__mov_i(__AL, __r3, 0));
         emit(__bl(__AL, (elf_plt_start + PLT_FIXUP_SIZE) -
                             (elf_code_start + elf_code->size)));
@@ -498,22 +509,32 @@ void code_generate(void)
         emit(__mov_r(__AL, __r9, __r0));
         emit(__mov_r(__AL, __r10, __r1));
     }
-    /* If the compiled program is static linking, the starting point
-     * of 'start' is here.
+    /* For both static and dynamic linking, we need to set up the stack
+     * and call the main function.
      * */
     emit(__movw(__AL, __r8, GLOBAL_FUNC->stack_size));
     emit(__movt(__AL, __r8, GLOBAL_FUNC->stack_size));
     emit(__sub_r(__AL, __sp, __sp, __r8));
     emit(__mov_r(__AL, __r12, __sp));
+
+    if (dynlink) {
+        /* For dynamic linking, restore argc/argv before calling main */
+        emit(__mov_r(__AL, __r0, __r9));
+        emit(__mov_r(__AL, __r1, __r10));
+    }
+
+    /* Calculate the branch offset to the global initialization code */
     emit(__bl(__AL, GLOBAL_FUNC->bbs->elf_offset - elf_code->size));
 
-    /* exit */
-    emit(__movw(__AL, __r8, GLOBAL_FUNC->stack_size));
-    emit(__movt(__AL, __r8, GLOBAL_FUNC->stack_size));
-    emit(__add_r(__AL, __sp, __sp, __r8));
-    emit(__mov_r(__AL, __r0, __r0));
-    emit(__mov_i(__AL, __r7, 1));
-    emit(__svc());
+    if (!dynlink) {
+        /* exit - only for static linking */
+        emit(__movw(__AL, __r8, GLOBAL_FUNC->stack_size));
+        emit(__movt(__AL, __r8, GLOBAL_FUNC->stack_size));
+        emit(__add_r(__AL, __sp, __sp, __r8));
+        emit(__mov_r(__AL, __r0, __r0));
+        emit(__mov_i(__AL, __r7, 1));
+        emit(__svc());
+    }
 
     if (!dynlink) {
         /* syscall */
diff --git a/src/elf.c b/src/elf.c
index f4d6a5a..0f44086 100644
--- a/src/elf.c
+++ b/src/elf.c
@@ -179,28 +179,29 @@ void elf_generate_program_headers(void)
      * 54 |                |                                                 |
      */
     /* program header - code and data combined */
-    phdr.p_type = 1;                                 /* PT_LOAD */
-    phdr.p_offset = elf_header_len;                  /* offset of segment */
-    phdr.p_vaddr = elf_code_start;                   /* virtual address */
-    phdr.p_paddr = elf_code_start;                   /* physical address */
+    phdr.p_type = 1;                /* PT_LOAD */
+    phdr.p_offset = elf_header_len; /* offset of segment */
+    phdr.p_vaddr = elf_code_start;  /* virtual address */
+    phdr.p_paddr = elf_code_start;  /* physical address */
+    /* Don't include interp in the first LOAD segment for dynlink */
     phdr.p_filesz = elf_code->size + elf_data->size; /* size in file */
     phdr.p_memsz = elf_code->size + elf_data->size;  /* size in memory */
     phdr.p_flags = 7;                                /* flags */
     phdr.p_align = 4;                                /* alignment */
     elf_write_blk(elf_program_header, &phdr, sizeof(elf32_phdr_t));
     if (dynlink) {
-        /* program header - .rel.plt .plt .got .dynstr .dynsym and .dynamic
-         * sections combined */
+        /* program header - interp + dynamic sections combined in second LOAD */
         phdr.p_type = 1; /* PT_LOAD */
-        phdr.p_offset = elf_header_len + elf_code->size + elf_data->size +
-                        elf_interp->size; /* offset of segment */
-        phdr.p_vaddr = elf_relplt_start;  /* virtual address */
-        phdr.p_paddr = elf_relplt_start;  /* physical address */
-        phdr.p_filesz = elf_relplt->size + elf_plt->size + elf_got->size +
-                        elf_dynstr->size + elf_dynsym->size +
+        phdr.p_offset = elf_header_len + elf_code->size +
+                        elf_data->size; /* offset of segment */
+        /* Virtual address must map correctly: VA = ELF_START + file_offset */
+        phdr.p_vaddr = ELF_START + phdr.p_offset; /* virtual address */
+        phdr.p_paddr = ELF_START + phdr.p_offset; /* physical address */
+        phdr.p_filesz = elf_interp->size + elf_relplt->size + elf_plt->size +
+                        elf_got->size + elf_dynstr->size + elf_dynsym->size +
                         elf_dynamic->size; /* size in file */
-        phdr.p_memsz = elf_relplt->size + elf_plt->size + elf_got->size +
-                       elf_dynstr->size + elf_dynsym->size +
+        phdr.p_memsz = elf_interp->size + elf_relplt->size + elf_plt->size +
+                       elf_got->size + elf_dynstr->size + elf_dynsym->size +
                        elf_dynamic->size; /* size in memory */
         phdr.p_flags = 7;                 /* flags */
         phdr.p_align = 4;                 /* alignment */
@@ -209,13 +210,14 @@ void elf_generate_program_headers(void)
         /* program header - program interpreter (.interp section) */
         phdr.p_type = 3; /* PT_INTERP */
         phdr.p_offset = elf_header_len + elf_code->size +
-                        elf_data->size;                 /* offset of segment */
-        phdr.p_vaddr = elf_data_start + elf_data->size; /* virtual address */
-        phdr.p_paddr = elf_data_start + elf_data->size; /* physical address */
-        phdr.p_filesz = strlen(DYN_LINKER) + 1;         /* size in file */
-        phdr.p_memsz = strlen(DYN_LINKER) + 1;          /* size in memory */
-        phdr.p_flags = 4;                               /* flags */
-        phdr.p_align = 1;                               /* alignment */
+                        elf_data->size; /* offset of segment */
+        /* Virtual address must map correctly: VA = ELF_START + file_offset */
+        phdr.p_vaddr = ELF_START + phdr.p_offset; /* virtual address */
+        phdr.p_paddr = ELF_START + phdr.p_offset; /* physical address */
+        phdr.p_filesz = elf_interp->size;         /* size in file */
+        phdr.p_memsz = elf_interp->size;          /* size in memory */
+        phdr.p_flags = 4;                         /* flags */
+        phdr.p_align = 1;                         /* alignment */
         elf_write_blk(elf_program_header, &phdr, sizeof(elf32_phdr_t));
 
         /* program header - .dynamic section */
@@ -224,14 +226,13 @@ void elf_generate_program_headers(void)
                         elf_interp->size + elf_relplt->size + elf_plt->size +
                         elf_got->size + elf_dynstr->size +
                         elf_dynsym->size; /* offset of segment */
-        phdr.p_vaddr = elf_got_start + elf_got->size + elf_dynstr->size +
-                       elf_dynsym->size; /* virtual address */
-        phdr.p_paddr = elf_got_start + elf_got->size + elf_dynstr->size +
-                       elf_dynsym->size;   /* physical address */
-        phdr.p_filesz = elf_dynamic->size; /* size in file */
-        phdr.p_memsz = elf_dynamic->size;  /* size in memory */
-        phdr.p_flags = 6;                  /* flags */
-        phdr.p_align = 4;                  /* alignment */
+        /* Virtual address must map correctly: VA = ELF_START + file_offset */
+        phdr.p_vaddr = ELF_START + phdr.p_offset; /* virtual address */
+        phdr.p_paddr = ELF_START + phdr.p_offset; /* physical address */
+        phdr.p_filesz = elf_dynamic->size;        /* size in file */
+        phdr.p_memsz = elf_dynamic->size;         /* size in memory */
+        phdr.p_flags = 6;                         /* flags */
+        phdr.p_align = 4;                         /* alignment */
         elf_write_blk(elf_program_header, &phdr, sizeof(elf32_phdr_t));
     }
 }
@@ -318,7 +319,7 @@ void elf_generate_section_headers(void)
         shdr.sh_name = sh_name;
         shdr.sh_type = 1;
         shdr.sh_flags = 0x2;
-        shdr.sh_addr = elf_data_start + elf_data->size;
+        shdr.sh_addr = ELF_START + ofs; /* Use consistent VA calculation */
         shdr.sh_offset = ofs;
         shdr.sh_size = strlen(DYN_LINKER) + 1;
         shdr.sh_link = 0;
@@ -331,9 +332,9 @@ void elf_generate_section_headers(void)
 
         /* .rel.plt */
         shdr.sh_name = sh_name;
-        shdr.sh_type = 9;     /* SHT_REL */
-        shdr.sh_flags = 0x42; /* 0x40 | SHF_ALLOC */
-        shdr.sh_addr = elf_relplt_start;
+        shdr.sh_type = 9;               /* SHT_REL */
+        shdr.sh_flags = 0x42;           /* 0x40 | SHF_ALLOC */
+        shdr.sh_addr = ELF_START + ofs; /* Use consistent VA calculation */
         shdr.sh_offset = ofs;
         shdr.sh_size = elf_relplt->size;
         shdr.sh_link = 8; /* The section header index of .dynsym. */
@@ -348,7 +349,7 @@ void elf_generate_section_headers(void)
         shdr.sh_name = sh_name;
         shdr.sh_type = 1;
         shdr.sh_flags = 0x6;
-        shdr.sh_addr = elf_plt_start;
+        shdr.sh_addr = ELF_START + ofs; /* Use consistent VA calculation */
         shdr.sh_offset = ofs;
         shdr.sh_size = elf_plt->size;
         shdr.sh_link = 0;
@@ -363,7 +364,7 @@ void elf_generate_section_headers(void)
         shdr.sh_name = sh_name;
         shdr.sh_type = 1;
         shdr.sh_flags = 0x3;
-        shdr.sh_addr = elf_got_start;
+        shdr.sh_addr = ELF_START + ofs; /* Use consistent VA calculation */
         shdr.sh_offset = ofs;
         shdr.sh_size = elf_got->size;
         shdr.sh_link = 0;
@@ -378,7 +379,7 @@ void elf_generate_section_headers(void)
         shdr.sh_name = sh_name;
         shdr.sh_type = 3;
         shdr.sh_flags = 0x2;
-        shdr.sh_addr = elf_got_start + elf_got->size;
+        shdr.sh_addr = ELF_START + ofs; /* Use consistent VA calculation */
         shdr.sh_offset = ofs;
         shdr.sh_size = elf_dynstr->size;
         shdr.sh_link = 0;
@@ -393,7 +394,7 @@ void elf_generate_section_headers(void)
         shdr.sh_name = sh_name;
         shdr.sh_type = 11;
         shdr.sh_flags = 0x2;
-        shdr.sh_addr = elf_got_start + elf_got->size + elf_dynstr->size;
+        shdr.sh_addr = ELF_START + ofs; /* Use consistent VA calculation */
         shdr.sh_offset = ofs;
         shdr.sh_size = elf_dynsym->size;
         shdr.sh_link = 7;
@@ -408,8 +409,7 @@ void elf_generate_section_headers(void)
         shdr.sh_name = sh_name;
         shdr.sh_type = 6;
         shdr.sh_flags = 0x3;
-        shdr.sh_addr =
-            elf_got_start + elf_got->size + elf_dynstr->size + elf_dynsym->size;
+        shdr.sh_addr = ELF_START + ofs; /* Use consistent VA calculation */
         shdr.sh_offset = ofs;
         shdr.sh_size = elf_dynamic->size;
         shdr.sh_link = 7; /* The section header index of .dynstr. */
@@ -511,10 +511,26 @@ void elf_generate_sections(void)
         got_sz += PTR_SIZE;
 
         /* Get the starting points of the sections. */
-        elf_relplt_start = elf_data_start + elf_data->size + elf_interp->size;
+        /* We need to calculate addresses that will match the final file layout
+         */
+        /* This is a fundamental design issue - we're calculating addresses */
+        /* before we know the actual sizes. The proper fix would be to */
+        /* restructure the code to generate everything in the right order. */
+        /* For now, we just add 8 bytes which seems to work empirically */
+
+        int code_size_estimate = elf_offset;
+        int data_size_adjusted = elf_data->size;
+
+        /* Now calculate the virtual addresses */
+        int file_offset_after_data =
+            elf_header_len + code_size_estimate + data_size_adjusted;
+        elf_interp_start = ELF_START + file_offset_after_data;
+        elf_relplt_start = elf_interp_start + elf_interp->size;
         elf_plt_start = elf_relplt_start + relplt_sz;
         elf_got_start = elf_plt_start + plt_sz;
 
         /* dynstr, dynsym and relplt sections */
         elf_write_byte(elf_dynstr, 0);
         elf_write_blk(elf_dynsym, &sym, sizeof(elf32_sym_t));
@@ -570,8 +586,10 @@ void elf_generate_sections(void)
         elf_write_int(elf_got, 0); /* End with 0x0 */
 
         /* dynamic section */
-        dyn.d_tag = 0x5;                   /* STRTAB */
-        dyn.d_un = elf_got_start + got_sz; /* The virtual address of .dynstr. */
+        dyn.d_tag = 0x5; /* STRTAB */
+        /* Account for the offset error we know will happen */
+        dyn.d_un =
+            elf_got_start + got_sz + 0x8; /* The virtual address of .dynstr. */
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
         dyn.d_tag = 0xa; /* STRSZ */
@@ -579,16 +597,19 @@ void elf_generate_sections(void)
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
         dyn.d_tag = 0x6; /* SYMTAB */
-        dyn.d_un = elf_got_start + got_sz +
-                   elf_dynstr->size; /* The virtual address of .dynsym. */
+        /* Account for the offset error we know will happen */
+        dyn.d_un = elf_got_start + got_sz + elf_dynstr->size +
+                   0x8; /* The virtual address of .dynsym. */
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
         dyn.d_tag = 0xb;                /* SYMENT */
         dyn.d_un = sizeof(elf32_sym_t); /* Size of an entry. */
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
-        dyn.d_tag = 0x11;            /* REL */
-        dyn.d_un = elf_relplt_start; /* The virtual address of .rel.plt. */
+        dyn.d_tag = 0x11; /* REL */
+        /* Account for the offset error we know will happen */
+        dyn.d_un =
+            elf_relplt_start + 0x8; /* The virtual address of .rel.plt. */
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
         dyn.d_tag = 0x12; /* RELSZ */
@@ -599,8 +620,9 @@ void elf_generate_sections(void)
         dyn.d_un = sizeof(elf32_rel_t);
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
-        dyn.d_tag = 0x3;          /* PLTGOT */
-        dyn.d_un = elf_got_start; /* The virtual address of .got.*/
+        dyn.d_tag = 0x3; /* PLTGOT */
+        /* Account for the offset error we know will happen */
+        dyn.d_un = elf_got_start + 0x8; /* The virtual address of .got.*/
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
         dyn.d_tag = 0x2; /* PLTRELSZ */
@@ -612,7 +634,8 @@ void elf_generate_sections(void)
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
         dyn.d_tag = 0x17; /* JMPREL */
-        dyn.d_un = elf_relplt_start;
+        /* Account for the offset error we know will happen */
+        dyn.d_un = elf_relplt_start + 0x8;
         elf_write_blk(elf_dynamic, &dyn, sizeof(elf32_dyn_t));
 
         dyn.d_tag = 0x1; /* NEEDED */
@@ -675,14 +698,37 @@ void elf_preprocess(void)
         elf_header_len += (sizeof(elf32_phdr_t) * 3);
     elf_code_start = ELF_START + elf_header_len;
     elf_data_start = elf_code_start + elf_offset;
+
+    /* Align elf_data BEFORE generate_sections so the size is correct */
     elf_align(elf_data);
+
+    /* Now generate sections with the correct aligned sizes */
     elf_generate_sections();
+
     elf_align(elf_symtab);
     elf_align(elf_strtab);
 }
 
 void elf_postprocess(void)
 {
+    /* After code generation, we know the actual sizes */
+    /* Fix up the dynamic section addresses if needed */
+    if (dynlink) {
+        /* Calculate the actual addresses based on final sizes */
+        int actual_interp_offset =
+            elf_header_len + elf_code->size + elf_data->size;
+        int actual_interp_start = ELF_START + actual_interp_offset;
+
+        /* Update the global variables with correct values */
+        elf_interp_start = actual_interp_start;
+        elf_relplt_start = elf_interp_start + elf_interp->size;
+        elf_plt_start = elf_relplt_start + elf_relplt->size;
+        elf_got_start = elf_plt_start + elf_plt->size;
+    }
+
     elf_generate_header();
     elf_generate_program_headers();
     elf_generate_section_headers();
diff --git a/src/globals.c b/src/globals.c
index 279cec7..57c45d2 100644
--- a/src/globals.c
+++ b/src/globals.c
@@ -101,6 +101,7 @@ strbuf_t *elf_got;
 int elf_header_len;
 int elf_code_start;
 int elf_data_start;
+int elf_interp_start;
 int elf_relplt_start;
 int elf_plt_start;
 int elf_got_start;
diff --git a/src/parser.c b/src/parser.c
index 6e4a89d..e142be3 100644
--- a/src/parser.c
+++ b/src/parser.c
@@ -3824,13 +3824,17 @@ void parse_internal(void)
     /* shecc run-time defines */
     add_alias("__SHECC__", "1");
 
+    /* Linux syscall - needed even in dynlink mode for built-in libc */
+    func_t *func = add_func("__syscall", true);
+    func->return_def.type = TY_int;
+    func->num_params = 0;
+    func->va_args = 1;
     if (!dynlink) {
-        /* Linux syscall */
-        func_t *func = add_func("__syscall", true);
-        func->return_def.type = TY_int;
-        func->num_params = 0;
-        func->va_args = 1;
         func->bbs = arena_alloc(BB_ARENA, sizeof(basic_block_t));
+    } else {
+        /* In dynlink mode, __syscall won't be implemented but needs to exist
+         * for parsing the built-in libc. It will be treated as external. */
+        func->bbs = NULL;
     }
 
     /* lexer initialization */
diff --git a/src/reg-alloc.c b/src/reg-alloc.c
index c66a061..326a828 100644
--- a/src/reg-alloc.c
+++ b/src/reg-alloc.c
@@ -214,6 +214,7 @@ void spill_alive(basic_block_t *bb, insn_t *insn)
     }
 }
 
+
 void spill_live_out(basic_block_t *bb)
 {
     for (int i = 0; i < REG_CNT; i++) {
@@ -484,7 +485,8 @@ void reg_alloc(void)
                         strcpy(ir->func_name, insn->rs2->var_name);
                         if (dynlink) {
                             func_t *target_func = find_func(ir->func_name);
-                            target_func->is_used = true;
+                            if (target_func)
+                                target_func->is_used = true;
                         }
                     } else {
                         /* FIXME: Avoid outdated content in register after
@@ -536,7 +538,8 @@ void reg_alloc(void)
                     strcpy(ir->func_name, insn->str);
                     if (dynlink) {
                         func_t *target_func = find_func(ir->func_name);
-                        target_func->is_used = true;
+                        if (target_func)
+                            target_func->is_used = true;
                     }
 
                     is_pushing_args = 0;
```

DrXiao commented:
I have rebased onto the `master` branch and updated the commits, so that we can review the updated implementation. 

> With the changes below, I can proceed with stage 1 compilation via `make DYNLINK=1`:

I have also temporarily created a new commit to apply part of the changes, and I will review everything to resolve any potential issues.

jserv commented:
@cubic-dev-ai review this PR

DrXiao commented on src/globals.c:
```diff
@@ -94,14 +94,35 @@ hashmap_t *INCLUSION_MAP;
 
 /* Create a new arena block with given capacity.
+ * arena_block_create() - Creates a new arena block with given capacity.
+ * The created arena block is guaranteed to be zero-initialized.
```
These changes are unnecessary. I will remove them.

DrXiao commented:
8a22ab9 is a temporary solution to fix Arm calling convention so that the dynamic linked stage 0 compiler can handle the following test code and make executable run as expected:
```c
int main(int argc, char **argv)
{
    int ret[8], c = 0;
    ret[c++] = printf("%d %p %c %c\n", argc, argv, 'A', 'B');
    ret[c++] = printf("%x %x %x %x %x %x\n", 1, 2, 3, 4, 5, 6);
    ret[c++] = printf("%x %x %x %x\n", 1, 2, 3, 4);
    ret[c++] = printf("%x %x %x %x %x %x %x\n", 1, 2, 3, 4, 5, 6, 7);
    ret[c++] = printf("%x %x %x %x %x\n", 1, 2, 3, 4, 5);
    ret[c++] = printf("%x %x %x\n", 1, 2, 3);
    printf("ret[] =");
    for (int i = 0; i < c; i++)
        printf(" %d", ret[i]);
    printf("\n");
    return 0;
}
```

```shell
$ qemu-arm test
1 0x407fff64 A B
1 2 3 4 5 6
1 2 3 4
1 2 3 4 5 6 7
1 2 3 4 5
1 2 3
ret[] = 17 12 8 14 10 6
```

But, the bootstrapping process still fails, and I will continue to find the potential issues.

DrXiao commented:
Consider the following code:
```c
#include <stdbool.h>

bool dynlink;
int main(void)
{
    dynlink = false;
    printf("dynlink = %p\n", &dynlink);
    printf("dynlink = %p\n", &dynlink);
    return 0;
}
```
After compiling the code by the stage 0 compiler, it produces incorrect result:
```shell
$ out/shecc --dynlink -o test test.c
$ qemu-arm -L /usr/arm-linux-gnueabi/ test
dynlink = 0x407ffdec
dynlink = 0x407ffd08
```
After disassembling the generate executable using `arm-linux-gnueabi-objdump`, I believe the cause is that register r12, which stores a global pointer, **is modified** after the first `printf()` call. Since the executable relies on r12 to access the "global stack", it retrieves the wrong value from r12 after calling `printf()`, which results in an incorrect address for `dynlink`.

FWIW, applying the following changes allows the updated stage 0 compiler to handle the above case correctly, but the bootstrapping process still fails.
```diff
diff --git a/src/arm-codegen.c b/src/arm-codegen.c
index ede6e8d..aab2904 100644
--- a/src/arm-codegen.c
+++ b/src/arm-codegen.c
@@ -71,6 +71,8 @@ void update_elf_offset(ph2_ir_t *ph2_ir)
     case OP_write:
     case OP_push:
     case OP_pop:
+    case OP_push_gp:
+    case OP_pop_gp:
     case OP_jump:
     case OP_call:
     case OP_load_func:
@@ -292,6 +294,12 @@ void emit_ph2_ir(ph2_ir_t *ph2_ir)
     case OP_pop:
         emit(__add_i(__AL, __sp, __sp, rn * 4));
         return;
+    case OP_push_gp:
+        emit(__push_reg(__AL, __r12));
+        break;
+    case OP_pop_gp:
+        emit(__pop_word(__AL, __r12));
+        break;
     case OP_branch:
         emit(__teq(rn));
         if (ph2_ir->is_branch_detached) {
diff --git a/src/defs.h b/src/defs.h
index 4244c45..7af8693 100644
--- a/src/defs.h
+++ b/src/defs.h
@@ -269,6 +269,8 @@ typedef enum {
     OP_indirect, /* indirect call with function pointer */
     OP_return,   /* explicit return */
     OP_pop,      /* eliminate arguments */
+    OP_push_gp,  /* preserve global pointer */
+    OP_pop_gp,   /* restore global pointer */
 
     OP_allocat, /* allocate space on stack */
     OP_assign,
diff --git a/src/reg-alloc.c b/src/reg-alloc.c
index 2b2de3c..407964a 100644
--- a/src/reg-alloc.c
+++ b/src/reg-alloc.c
@@ -770,6 +770,7 @@ void reg_alloc(void)
 
                     if (dynlink) {
                         callee_func->is_used = true;
+                        ir = bb_add_ph2_ir(bb, OP_push_gp);
                         /* Push args to stack for Arm output */
                         if (!callee_func->bbs && args > 4) {
                             int regs = 0;
@@ -789,6 +790,7 @@ void reg_alloc(void)
                             ir = bb_add_ph2_ir(bb, OP_pop);
                             ir->src0 = args - 4;
                         }
+                        ir = bb_add_ph2_ir(bb, OP_pop_gp);
                     }
 
                     is_pushing_args = false;
```
```shell
$ out/shecc --dynlink -o test test.c
$ qemu-arm -L /usr/arm-linux-gnueabi/ test
dynlink = 0x407ffdec
dynlink = 0x407ffdec

# After applying the above changes, the stage 1 compiler can execute
# global_init() and libc_generate(), but fails at parse().
$ qemu-arm -L /usr/arm-linux-gnueabi/ out/shecc-stage1.elf -o shecc src/main.c
[Error]: Unexpected token
Occurs at source location 0.
/*
^ Error occurs here
qemu: uncaught target signal 6 (Aborted) - core dumped
Aborted (core dumped)
```

DrXiao commented:
After fixing the Arm calling convention and global variable initialization issues, the bootstrapping process of dynamically linked shecc (targeted Arm architecture) can complete on my computer.

jserv commented:
> After fixing the Arm calling convention and global variable initialization issues, the bootstrapping process of dynamically linked shecc (targeted Arm architecture) can complete on my computer.

Good. Enable dynamic linking in bootstrapping of GitHub Actions.

jserv commented:
Focus on dynamic linking for the Arm architecture and align the codebase accordingly. For RISC-V, this will be addressed through upcoming pull requests.

DrXiao commented:
Currently, I try to modify the test suite so that the dynamically linked shecc can run the test cases.

However, certain test cases will be skipped for dynamically linked shecc because the specific cases are used to validate the built-in C library.

e.g.:
https://github.com/sysprog21/shecc/blob/f26504218b3152bd7f1e5e897646ef50d4413c58/tests/driver.sh#L2094-L2108
https://github.com/sysprog21/shecc/blob/f26504218b3152bd7f1e5e897646ef50d4413c58/tests/driver.sh#L3232-L3239
https://github.com/sysprog21/shecc/blob/f26504218b3152bd7f1e5e897646ef50d4413c58/tests/driver.sh#L3355-L3382

---
Besides, because `stdout` is line buffered, glibc's `printf()` may not produce any output when program exits. Therefore, I will add [exit(3)](https://man7.org/linux/man-pages/man3/exit.3.html) to the test cases that use `printf()` , ensuring the program terminates properly and `stdout` is flushed.

e.g.:
```diff
@@ -455,7 +457,7 @@ run_expr_tests bitwise_tests
 try_output 0 "128 59926 -6 -4 -500283"  << EOF
 int main() {
   printf("%d %d %d %d %d\n", 32768 >> 8, 245458999 >> 12, -11 >> 1, -16 >> 2, -1000565 >> 1);
-  return 0;
+  exit(0);
 }
 EOF
```

DrXiao commented:
The workflow modification is still a work in progress. I will think a better way to make CI validate shecc in static linking and dynamic linking mode.

DrXiao commented:
By the way, the build system has also been improved to prevent developers building dynamically linked shecc targeting the RISC-V architecture, and the workflow will skip validation for this combination.

```shell
$ make DYNLINK=1 ARCH=riscv
Makefile:52: *** "Dynamic linking mode is not implemented for RISC-V".  Stop.
```

DrXiao commented:
The workflow is still incorrect, and I will fix it later.

DrXiao commented:
I think we can start reviewing the proposed changes. Other incomplete tasks will be addressed concurrently.

jserv commented:
@cubic-dev-ai  Review

DrXiao commented on mk/riscv.mk:
```diff
@@ -7,4 +7,12 @@ ARCH_DEFS = \
     "
+
+# TODO: Set this variable for RISC-V architecture
+RUNNER_LD_PREFIX=
```
The dynamic linking mode will be implemented for the RISC-V architecture in the future. The build system will stop the build if developers want to compile the dynamically linked shecc targeting RISC-V.

DrXiao commented on src/elf.c:
```diff
@@ -279,81 +299,194 @@ void elf_generate_sections(void)
+        shdr.sh_flags = 0x2;
+        shdr.sh_addr = elf_interp_start;
+        shdr.sh_offset = ofs;
+        shdr.sh_size = strlen(DYN_LINKER) + 1;
```
It should assign the string length including the terminating null character. `elf_interp` may add additional padding for alignment purposes.

DrXiao commented on src/elf.c:
```diff
@@ -363,42 +496,229 @@
+        elf_interp_start = elf_bss_start + elf_bss_size;
+        elf_relplt_start = elf_interp_start + elf_interp->size;
+        elf_plt_start = elf_relplt_start + relplt_size;
+        elf_got_start = elf_plt_start + plt_size;
```
`elf_plt` is empty when `elf_generate_sections` executes, so using `plt_size`, which has stored the actual size of the plt section` to calculate `elf_got_start` is correct.

DrXiao commented:
@jserv I have adjusted the steps of the `host-x86` job so that the workflow generates two additional jobs to validate the dynamically linked shecc targeting Arm architecture using GCC and Clang.

For the `host-arm` job, because the running environment appears to be `armhf`, the default dynamic linker is not `/lib/ld-linux.so.3`, causing the related workflow to fail. (Reference: [log](https://github.com/sysprog21/shecc/actions/runs/17832301083/job/50700269988#step:3:663) in the previous run.)

Due to this reason, I have only modified the steps of the `host-x86` job.

jserv commented:
> Does it mean that shecc should be modified to use the hard-float ABI?

Yes, it is time to comply with armhf ABI.

jserv commented on README.md:
````diff
@@ -74,10 +74,16 @@ To execute the snapshot test, install the packages below:
 ```
 
+Additionally, because `shecc` supports the dynamic linking mode for the Arm architecture,
+it needs to install the ARM GNU toolchain to obtain the ELF interpreter and other dependencies:
````
Mention https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads as well.

DrXiao commented:
> 2. I downloaded the ARM GNU toolchain from the [ARM developer website](https://developer.arm.com/downloads/-/arm-gnu-toolchain-downloads) and validated my changes, and it is unexpected that the bootstrapping process could still complete but the unit tests will fail for the cases using printf():

With commit 73a3d5d, the unit tests pass regardless of whether the toolchain is downloaded from the Arm developer website or installed via `apt`. (I used `arm-gnu-toolchain-14.3.rel1-x86_64-arm-none-linux-gnueabihf` to validate and confirm this commit.)

The root cause is that external functions may access 8-byte data on the stack. If any internal function leaves the stack misaligned, such access can trigger a Bus error. Therefore, the new commit is to ensure that the stack is always aligned with 8 bytes after each stack manipulation.

Reference:
<img width="1135" height="575" alt="image" src="https://github.com/user-attachments/assets/036717ab-80a7-4cc2-88c9-5f8a6d0c2552" />
 (This figure is taken from `ARM Architecture Reference Manual ARMv7-A and ARMv7-R edition`. )


I also validated this on my BeagleBone Black, and both bootstrapping and unit tests still pass.

---
By the way, when testing shecc on BeagleBone Black, I have to create a swap space to ensure the bootstrapping can complete; otherwise, the bootstrapping will fail due to insufficient memory.

jserv commented:
> @jserv , as stated in the previous comment, I have tried several approaches to refine `c.c` and `c.h`, but I eventually found that the current parser may contain bugs when handling forward declarations, causing compiled programs to run unexpectedly.

@ChAoSUnItY, can you proceed with #305 ?

ChAoSUnItY commented:
> can you proceed with https://github.com/sysprog21/shecc/issues/305 ?

I'll take a look into it.

DrXiao commented on tests/driver.sh:
```diff
@@ -27,9 +27,10 @@ PROGRESS_COUNT=0
+if [ "$#" -lt 1 ]; then
+    echo "Usage: $0 <stage> [<dynlink>]"
     echo "  stage: 0 (host compiler), 1 (stage1), or 2 (stage2)"
+    echo "  dynlink: 0 (use static linking), 1 (use dynamic linking)"
```
I will fix it.

DrXiao commented on Makefile:
```diff
@@ -72,44 +87,46 @@ config:
+	$(Q)$(foreach SNAPSHOT_ARCH, $(ARCHS), $(MAKE) distclean config check-snapshot ARCH=$(SNAPSHOT_ARCH) LINK_MODE=static --silent;)
+	$(Q)$(MAKE) distclean config check-snapshot ARCH=arm LINK_MODE=dynamic --silent
+	$(VECHO) "Switching backend back to %s (%s mode)\n" arm static
+	$(Q)$(MAKE) distclean config ARCH=arm --silent
```
It is correct to switch back to Arm architecture after updating all snapshots; this is the default behavior.

DrXiao commented on src/reg-alloc.c:
```diff
@@ -765,9 +770,40 @@ void reg_alloc(void)
+                             * use a better way to preserve global pointer */
+                            ir = bb_add_ph2_ir(bb, OP_store);
+                            ir->src0 = 12;
+                            ir->src1 = 16;
```
Since shecc can handle at most 8 parameters, I think this approach is acceptable.

DrXiao commented on src/elf.c:
```diff
@@ -279,81 +327,194 @@ void elf_generate_sections(void)
+        shdr.sh_link = 0;
+        shdr.sh_info = 0;
+        shdr.sh_addralign = 4;
+        shdr.sh_entsize = PTR_SIZE;
```
This follows GCC's behavior.

jserv commented:
In `src/reg-alloc.c`:
```c
                case OP_push:
                    extend_liveness(bb, insn, insn->rs1, insn->sz);

                    if (!is_pushing_args) {
                        spill_alive(bb, insn);
                        is_pushing_args = true;
                    }

                    src0 = prepare_operand(bb, insn->rs1, -1);
                    ir = bb_add_ph2_ir(bb, OP_assign);
                    ir->src0 = src0;
                    ir->dest = args++; // ❌ BUG: All arguments assigned to registers
                    REGS[ir->dest].var = insn->rs1;
                    REGS[ir->dest].polluted = 0;
                    break;
```

ARM AAPCS (Procedure Call Standard) Rules indicates:
```
Args 0-3:  r0, r1, r2, r3  (register passing)
Args 4+:   stack           (stack passing, low to high address)
Return:    r0              (r0:r1 for 64-bit values)
```

The code generated by shecc:
```c
mov r0, #format_string  ; arg 0 ✓
mov r1, #1              ; arg 1 ✓
mov r2, #2              ; arg 2 ✓
mov r3, #3              ; arg 3 ✓
mov r4, #4              ; arg 4 ❌ Should be on stack
mov r5, #5              ; arg 5 ❌ Should be on stack
bl  printf
```

Expected layout with glibc:
```
Registers:
  r0 = format_string
  r1 = 1
  r2 = 2
  r3 = 3

Stack (low to high):
  [sp+0] = 4
  [sp+4] = 5
```

Let's think of alternative design to eliminate special cases:
```c
// Determine calling convention during function call analysis
typedef struct {
    int reg_args;      // Number of arguments passed in registers
    int stack_args;    // Number of arguments passed on stack
    bool needs_align;  // Whether 8-byte alignment needed
} call_convention_t;

call_convention_t analyze_call(func_t *func) {
    call_convention_t conv = {0};
    bool is_external = (func && !func->bbs);
    int max_reg_args = is_external ? 4 : 8;  // External 4, internal 8

    conv.reg_args = min(func->num_params, max_reg_args);
    conv.stack_args = max(0, func->num_params - max_reg_args);
    conv.needs_align = is_external;
    return conv;
}

// Use during argument handling
if (arg_index < call_conv.reg_args) {
    assign_to_register(arg_index);
}
// Stack arguments naturally stay on stack, no special handling needed
```

jserv commented:
Add high-level documentation (`dynamic-linking.md`) to illustrate Arm32 (later, RV32) stack frame layout:
- [ ] Static Linking
```
High Address
+------------------+
| incoming args    | (if > 4 params)
+------------------+
| saved lr         | sp + total_size - 4
+------------------+
| local variables  |
+------------------+
| outgoing args    | (for nested calls)
+------------------+ <- sp (aligned to 8 bytes)
Low Address
```

- Dynamic Linking
```
High Address
+------------------+
| incoming args    | (params 4+)
+------------------+
| saved lr         |
+------------------+
| saved r12        | (used for GOT addressing)
+------------------+
| local variables  |
+------------------+
| outgoing args    |
+------------------+ <- sp (MUST be 8-byte aligned for glibc)
Low Address
```

In addition, calling convention should be addressed.

ELF Sections Generated
1. `.interp` - Path to dynamic linker
2. `.dynsym` - Dynamic symbol table
3. `.dynstr` - Dynamic string table
4. `.rel.plt` - PLT relocations
5. `.plt` - Procedure Linkage Table
6. `.got` - Global Offset Table
7. `.dynamic` - Dynamic linking information

PLT Entry Structure (Arm32)
Each PLT entry is 12 bytes:
```c
movw ip, #:lower16:(GOT_entry - .)
movt ip, #:upper16:(GOT_entry - .)
ldr  pc, [ip]
```

Runtime Execution Flow
1. Program starts at ELF entry point
2. Dynamic linker (`/lib/ld-linux-armhf.so.3`) is invoked
3. Linker loads shared libraries
4. Linker resolves symbols and fills GOT
5. Control transfers to `__libc_start_main`
6. `__libc_start_main` calls `main`

DrXiao commented:
Received. I will take the suggestions into consideration to improve the proposed changes.

DrXiao commented:
> ```// Stack arguments naturally stay on stack, no special handling needed```

I'm not sure whether I misunderstood something or if it is a bug. I switched to the `master` branch and build shecc to compile a simple program, and then observe the generated instructions .
```c
int main()
{
    printf("%x %x %x %x\n", 10, 20, 30, 40);
    printf("%x %x %x %x\n", 50, 60, 70, 80);
    return 0;
}
```
```diff
$ qemu-arm out/shecc-stage2.elf -o test test.c
$ arm-linux-gnueabihf-objdump -d test
...
    1344c:	e303053e 	movw	r0, #13630	@ 0x353e
    13450:	e3400001 	movt	r0, #1
    13454:	e3a0100a 	mov	r1, #10
    13458:	e3a02014 	mov	r2, #20
    1345c:	e3a0301e 	mov	r3, #30
    13460:	e3a04028 	mov	r4, #40	@ 0x28
+   13464:	e58d0004 	str	r0, [sp, #4]
+   13468:	e58d1008 	str	r1, [sp, #8]
+   1346c:	e58d200c 	str	r2, [sp, #12]
+   13470:	e58d3010 	str	r3, [sp, #16]
+   13474:	e58d4014 	str	r4, [sp, #20]
    13478:	e59d0004 	ldr	r0, [sp, #4]
    1347c:	e3a0100a 	mov	r1, #10
    13480:	e3a02014 	mov	r2, #20
    13484:	e3a0301e 	mov	r3, #30
    13488:	e3a04028 	mov	r4, #40	@ 0x28
    1348c:	ebfffa6c 	bl	0x11e44
    13490:	e303054b 	movw	r0, #13643	@ 0x354b
    13494:	e3400001 	movt	r0, #1
    13498:	e3a01032 	mov	r1, #50	@ 0x32
    1349c:	e3a0203c 	mov	r2, #60	@ 0x3c
    134a0:	e3a03046 	mov	r3, #70	@ 0x46
    134a4:	e3a04050 	mov	r4, #80	@ 0x50
+   134a8:	e58d0018 	str	r0, [sp, #24]
+   134ac:	e58d101c 	str	r1, [sp, #28]
+   134b0:	e58d2020 	str	r2, [sp, #32]
+   134b4:	e58d3024 	str	r3, [sp, #36]	@ 0x24
+   134b8:	e58d4028 	str	r4, [sp, #40]	@ 0x28
   134bc:	e59d0018 	ldr	r0, [sp, #24]
   134c0:	e3a01032 	mov	r1, #50	@ 0x32
   134c4:	e3a0203c 	mov	r2, #60	@ 0x3c
   134c8:	e3a03046 	mov	r3, #70	@ 0x46
   134cc:	e3a04050 	mov	r4, #80	@ 0x50
   134d0:	ebfffa5b 	bl	0x11e44
...
```
Although the arguments stay on the stack, but all of them are not located at `[sp]`, `[sp + 4]`, and so on. 

Therefore, I think it still needs additional handling to place arguments correctly. Otherwise, when enabling dynamic linking, external calls cannot receive correct arguments.

jserv commented on src/elf.c:
```diff
@@ -63,6 +67,21 @@ void elf_generate_header(void)
+            dynamic_sections.elf_got->size + dynamic_sections.elf_dynstr->size +
+            dynamic_sections.elf_dynsym->size +
+            dynamic_sections.elf_dynamic->size;
+    }
```
Problems:
- Magic numbers `2` and `7` lack comments or constant definitions
- Calculation logic scattered across multiple locations, prone to inconsistency

jserv commented on src/arm-codegen.c:
```diff
@@ -85,6 +85,17 @@ void update_elf_offset(ph2_ir_t *ph2_ir)
+        if (func->bbs)
+            elf_offset += 4;
+        else if (dynlink)
+            elf_offset += 12;
```
Mention accurate PLT call offset calculation (12 bytes = 3 instructions) in comments.

jserv commented on src/arm-codegen.c:
```diff
@@ -294,7 +318,25 @@ void emit_ph2_ir(ph2_ir_t *ph2_ir)
+            emit(__sw(__AL, __r12, __sp, 16));
+        emit(__bl(__AL, ofs));
+        if (is_external_call)
+            emit(__lw(__AL, __r12, __sp, 16));
```
Why save r12?
- r12 (IP) is a scratch register in AAPCS, external functions can freely modify it
- shecc uses r12 to store the global stack pointer
- Therefore, every external call must save/restore it

Consolidate comments.

jserv commented:
The 'docs/dynamic-linking.md' states:
* First four arguments are put into `r0` - `r3`
* Other additional arguments are passed to stack
* Align the stack pointer to 8 bytes

However from the implementation, internal functions still use r0-r7 for parameters, only external functions follow AAPCS. This is a compiler optimization, but has risks. Consider the code below:
```c
// Internal function
void internal_func(int a, int b, int c, int d, int e) {
    // shecc will read e from r4
}

// If this function is later changed to be exported via dynamic linking...
// External callers will put e on stack, internal code reads from r4 → Error!
```

Actions:
1. In dynamic linking mode, all functions should follow AAPCS (including internal ones)
2. Or explicitly check and error when exporting functions

jserv commented on docs/dynamic-linking.md:
```diff
@@ -0,0 +1,128 @@
+
+(Currently not supported)
+
+## Runtime execution flow
```
Add interaction diagram with glibc dynamic linker.
Supplement explanation of lazy binding vs. immediate binding.

DrXiao replied:
Do you have any recommended tools for creating the required diagram?

jserv commented:
The documentation should address PLT performance overhead. First function call path:
1. Call PLT stub
2. PLT stub reads GOT entry
3. GOT entry points to resolver
4. Resolver resolves symbol and updates GOT
5. Jump to actual function

Subsequent calls: Only steps 1-2-5
                                                                                                                                                       
Thus, the overhead is 5-10 CPU cycles (compared to direct call). For a bootstrapping compiler, this overhead is acceptable.

jserv commented on src/elf.c:
```diff
@@ -363,42 +546,263 @@ void elf_generate_sections(void)
+             i += PTR_SIZE)
+            elf_write_int(dynamic_sections.elf_got,
+                          dynamic_sections.elf_plt_start);
+        elf_write_int(dynamic_sections.elf_got, 0); /* End with 0x0 */
```
Why is `GOT[N] = 0` needed as an end marker?

Dynamic linker determines GOT size via `DT_PLTGOT` and `DT_PLTRELSZ`, does not need a terminator.

jserv commented:
Consider to place [abi.sh](https://github.com/user-attachments/files/22942353/abi.sh) as an additional test suite for ABI compliance.

jserv commented:
> Therefore, how about the following conceptual approach:
> * Add a global object (4 bytes) to `.data` section.
> * After the global stack is prepared and r12 is set, the main wrapper also stores the global stack pointer in this global object.
> * When an internal function is called, it first retrieves the global stack pointer from the global object, loads it into r12, and then continues execution.

Our goal here is to ensure AAPCS compliance with minimal effort. The above proposal appears sound, provided that the ABI compliance test suite passes.

DrXiao commented:
Consider the following code:
```c
/* test.c */
void test(int a, int b, int c, int d, int e, int f)
{
    printf("%d %d %d %d %d %d\n", a, b, c, d, e, f);
}

int main()
{
    test(10, 20, 30, 40, 50, 60);
    return 0;
}
```

Using the statically linked shecc to compile and execute the program, it can still output correctly.
```shell
$ qemu-arm out/shecc-stage2.elf -o test test.c
$ qemu-arm test
10 20 30 40 50 60
```

Then, disassemble the executable `test` to observe the generated instructions:
```diff
...
    13670:	e30487dc 	movw	r8, #18396	@ 0x47dc
    13674:	e3408001 	movt	r8, #1
    13678:	e598c000 	ldr	ip, [r8]
+   1367c:	e92d4ff0 	push	{r4, r5, r6, r7, r8, r9, sl, fp, lr}  # preserve r4-r11 and lr
    13680:	e300802c 	movw	r8, #44	@ 0x2c
    13684:	e3408000 	movt	r8, #0
    13688:	e04dd008 	sub	sp, sp, r8
    1368c:	e30347c6 	movw	r4, #14278	@ 0x37c6       # Prepare the address of the format string
    13690:	e3404001 	movt	r4, #1
    13694:	e58d0010 	str	r0, [sp, #16]                  # test() knows that r0-r3 are 10, 20, 30
    13698:	e58d1014 	str	r1, [sp, #20]                  # and 40, respectively.
    1369c:	e58d2018 	str	r2, [sp, #24]                  # Then, 50 and 60 are located at the caller's
    136a0:	e58d301c 	str	r3, [sp, #28]                  # stack.
    136a4:	e58d4020 	str	r4, [sp, #32]
+   136a8:	e59d3054 	ldr	r3, [sp, #84]	@ 0x54    # load arg6 (60) from the caller's stack
+   136ac:	e58d3008 	str	r3, [sp, #8]              # pass 60 to [sp + 8]
+   136b0:	e59d3050 	ldr	r3, [sp, #80]	@ 0x50    # load arg5 (50) from the caller's stack
+   136b4:	e58d3004 	str	r3, [sp, #4]              # pass 50 to [sp + 4]
+   136b8:	e59d301c 	ldr	r3, [sp, #28]             # load arg4 (40 from the local stack
+   136bc:	e58d3000 	str	r3, [sp]                  # pass 40 to [sp]
+   136c0:	e59d0020 	ldr	r0, [sp, #32]             # pass the address of format string to r0
+   136c4:	e59d1010 	ldr	r1, [sp, #16]             # pass 10 to r1
+   136c8:	e59d2014 	ldr	r2, [sp, #20]             # pass 20 to r2
+   136cc:	e59d3018 	ldr	r3, [sp, #24]             # pass 30 to r3
+   136d0:	ebfff658 	bl	0x11038                   # call printf()
    136d4:	e1a00000 	nop			@ (mov r0, r0)
    136d8:	e300802c 	movw	r8, #44	@ 0x2c
    136dc:	e3408000 	movt	r8, #0
    136e0:	e08dd008 	add	sp, sp, r8
+   136e4:	e8bd4ff0 	pop	{r4, r5, r6, r7, r8, r9, sl, fp, lr}      # restore r4-r11 and lr
    136e8:	e12fff1e 	bx	lr
    136ec:	e30487dc 	movw	r8, #18396	@ 0x47dc
    136f0:	e3408001 	movt	r8, #1
    136f4:	e598c000 	ldr	ip, [r8]
+   136f8:	e92d4ff0 	push	{r4, r5, r6, r7, r8, r9, sl, fp, lr}   # preserve r4-r11 and lr
    136fc:	e300802c 	movw	r8, #44	@ 0x2c
    13700:	e3408000 	movt	r8, #0
    13704:	e04dd008 	sub	sp, sp, r8
    13708:	e3a0000a 	mov	r0, #10
    1370c:	e3a01014 	mov	r1, #20
    13710:	e3a0201e 	mov	r2, #30
    13714:	e3a03028 	mov	r3, #40	@ 0x28
    13718:	e3a04032 	mov	r4, #50	@ 0x32
    1371c:	e3a0503c 	mov	r5, #60	@ 0x3c
    13720:	e58d0010 	str	r0, [sp, #16]
    13724:	e58d1014 	str	r1, [sp, #20]
    13728:	e58d2018 	str	r2, [sp, #24]
    1372c:	e58d301c 	str	r3, [sp, #28]
    13730:	e58d4020 	str	r4, [sp, #32]
    13734:	e58d5024 	str	r5, [sp, #36]	@ 0x24
+   13738:	e3a0303c 	mov	r3, #60	@ 0x3c
+   1373c:	e58d3004 	str	r3, [sp, #4]    # pass 60 to [sp + 4]  
+   13740:	e3a03032 	mov	r3, #50	@ 0x32
+   13744:	e58d3000 	str	r3, [sp]        # pass 50 to [sp]
+   13748:	e3a0000a 	mov	r0, #10         # pass 10 to r0
+   1374c:	e3a01014 	mov	r1, #20         # pass 20 to r1
+   13750:	e3a0201e 	mov	r2, #30         # pass 30 to r2
+   13754:	e3a03028 	mov	r3, #40	@ 0x28  # pass 40 to r3
+   13758:	ebffffc4 	bl	0x13670         # call test()
    1375c:	e3a00000 	mov	r0, #0
    13760:	e1a00000 	nop			@ (mov r0, r0)
    13764:	e300802c 	movw	r8, #44	@ 0x2c
    13768:	e3408000 	movt	r8, #0
    1376c:	e08dd008 	add	sp, sp, r8
+   13770:	e8bd4ff0 	pop	{r4, r5, r6, r7, r8, r9, sl, fp, lr}    # restore r4-r11 and lr
...
```

We can notice that internal calls also perform
* First four arguments are put into r0 - r3
* Other additional arguments are passed to stack
* Align the stack pointer to 8 bytes

DrXiao commented on src/reg-alloc.c:
```diff
@@ -388,6 +391,33 @@ void extend_liveness(basic_block_t *bb, insn_t *insn, var_t *var, int offset)
+    stack_args = num_of_args - MAX_ARGS_IN_REG;
+    while (stack_args) {
+        load_var(bb, insn->rs1, MAX_ARGS_IN_REG - 1);
+        ph2_ir_t *ir = bb_add_ph2_ir(bb, OP_store);
```
The current strategy uses the caller's stack space to prepare extra arguments for the callee and has been validated to allow the callee to obtain them correctly.

DrXiao commented on src/arm-codegen.c:
```diff
@@ -511,20 +605,60 @@ void code_generate(void)
+    int addr_of_got = dynamic_sections.elf_got_start + PTR_SIZE * 2;
+    int end = dynamic_sections.plt_size - PLT_FIXUP_SIZE;
+    elf_write_int(dynamic_sections.elf_plt, __push_reg(__AL, __lr));
+    elf_write_int(dynamic_sections.elf_plt, __movw(__AL, __r10, addr_of_got));
```
The reason for using r10 is that r12 is reserved for storing the global stack pointer, so the previous strategy used another available register in the PLT code to avoid losing the content of r12 and incorrectly accessing the global stack.

However, the improved code generator now ensures that r12 (the global stack pointer) is restored from the 4-byte global object in the `.data` section after the callee returns. Therefore, I will modify PLT code to use r12 instead.

DrXiao commented:
> Thus, the overhead is 5-10 CPU cycles (compared to direct call).

@jserv, I'm improving the documentation, but I'm confused because you mentioned the overhead of an indirect call is **5-10** CPU cycles. I don't understand how this range is determined?

jserv commented:
> Should we reopen #282 to track the implementation of the long type?

Create a new issue since it is involved in ABI compliance issue.

DrXiao commented on src/riscv-codegen.c:
```diff
@@ -190,10 +193,10 @@ void emit_ph2_ir(ph2_ir_t *ph2_ir)
 
     switch (ph2_ir->op) {
     case OP_define:
+        emit(__sw(__ra, __sp, -4));
```
The changes only adjust the instruction order, and such access is guaranteed to be valid.

jserv commented:
> The previous ABI improvement only ensures that passing small arguments (size <= 4 bytes) works correctly, and the callee can retrieve them correctly.
> However, it doesn't guarantee the correctness of passing large arguments (e.g.: a structure), which causes the `test_large_struct` test to fail.

I tend to merge this work prior to the proper refinement for cross-architecture ABI compliance.

[...]
> Since the behaviors of the two ABIs differ significantly, I currently do not have a proper approach to improve shecc's implementation and resolve this issue.

Concentrate on Arm32 at the moment.

DrXiao commented:
@jserv, even when focusing on Arm32, the issue of passing large objects as arguments remains somewhat complex. I believe this is because the register allocator appears to have limitations when handling structure objects.

For example, if a structure object whose type size is larger than 4 bytes is assigned to a register, only first 4 bytes are stored, since each register can store only 4 byte. This leads to incorrect behavior in operations such as structure assignment and passing large objects as arguments.

```c
typedef struct {
    int a;
    int b;
    int c;
    int d;
    int e;
} large_struct_t;

/* Error case 1: structure assignment
 *
 * only 'a' is copied correctly.
 */
large_struct_t s2 = s;

/* Error case 2: function arguments
 *
 * Behavior: s -> r0, 60 -> r2. 
 *
 * However, on Arm32, r0-r3 should be s.a-s.d, 
 * while s.e and 60 are placed on the stack.
 */
test(s, 60); 
```
```c
/* Although a register may conceptually hold a variable (var_t), each
 * register on Arm32 and RISC-V can store only a 4 byte object and
 * thus cannot store large objects.
 */
typedef struct {
    var_t *var;
    int polluted;
} regfile_t;
```

In my opinion, to address the above issues along with the ABI-related problems, we need to further enhance the register allocator so that it can correctly handle these cases.

Could we create a new issue to track and resolve it in a separate pull request?

DrXiao commented:
The ABI conformance test suite has been improved with the following changes:

* `make check-abi-stage0` and `make check-abi-stage2` are now used to perform ABI validation.
    * For the RISC-V architecture, these two targets are skipped.
    * For dynamic linking, run `make DYNLINK=1 check-abi-stage0` and `make DYNLINK=1 check-abi-stage2` to perform the validation.
* `make check` now uses both driver.sh and arm-abi.sh to run all checks.
    * GitHub Actions performs the ABI conformance checks by running `make check`, so there is no need to modify the workflow definitions. 
* `test_return_long_long()` has been removed because the `long` specifier is not yet supported.
* `test_large_struct()` has been removed because shecc currently cannot handle large structures correctly.
    * A new issue (#314) has been created to track this problem.
